test-data/unit/check-classes.test

```diff
@@ -2065,3 +2065,167 @@ class A: pass
+# reveal_type(B.copy(B()))
+
+
+[case testSelfTypeReturn]
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+def _type(self: T) -> Type[T]: ...
+
+class A:
+    def copy(self: T) -> T:
+        if B():
+            return A()
+        elif A():
+            return B()
+        return _type(self)()
+
+class B(A):
+    pass
+
+[out]
+main: note: In member "copy" of class "A":
+main:10: error: Incompatible return value type (got "A", expected "T")
+main:12: error: Incompatible return value type (got "B", expected "T")
+
+[case testSelfTypeClass]
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+class A:
+    @classmethod
+    def new(cls: Type[T]) -> T:
+        return cls()
+
+class B(A):
+    pass
+
+reveal_type(A.new)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(B.new)  # E: Revealed type is 'def () -> __main__.B'
+reveal_type(A.new())  # E: Revealed type is '__main__.A'
+reveal_type(B.new())  # E: Revealed type is '__main__.B'
+
+[builtins fixtures/classmethod.pyi]
+
+[case testSelfTypeOverride]
+from typing import TypeVar, cast
+ 
+T = TypeVar('T')
+ 
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    pass
+
+class C(A):
+    def copy(self: T) -> T: ...
+
+reveal_type(C().copy)  # E: Revealed type is 'def () -> __main__.C'
+reveal_type(C().copy())  # E: Revealed type is '__main__.C'
+reveal_type(cast(A, C()).copy)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(cast(A, C()).copy())  # E: Revealed type is '__main__.A' 
+
+[case testSelfTypeSuper]
+from typing import TypeVar, cast
+ 
+T = TypeVar('T')
+ 
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    def copy(self: T) -> T:
+        reveal_type(self)  # E: Revealed type is 'T`-1'
+        ## TODO:
+        ## super() dispatch is A.copy, but the type binding should be T=B
+        # reveal_type(super().copy())  # Revealed type is 'T`-1'
+        # return super().copy()
+        return cast(T, super().copy())
+
+[out]
+main: note: In member "copy" of class "B":
+
+[case testSelfTypeRecursiveBinding]
+from typing import TypeVar, Callable, Type
+ 
+T = TypeVar('T')
+class A:
+    def copy(self: T, factory: Callable[[T], T]) -> T:
+        return factory(self)
+
+    @classmethod
+    def new(cls: Type[T], factory: Callable[[T], T]) -> T:
+        return cls()
+
+class B(A):
+    pass
+
+reveal_type(A().copy) # E: Revealed type is 'def (factory: def (__main__.A) -> __main__.A) -> __main__.A'
+reveal_type(B().copy) # E: Revealed type is 'def (factory: def (__main__.B) -> __main__.B) -> __main__.B'
+reveal_type(A.new) # E: Revealed type is 'def (factory: def (__main__.A) -> __main__.A) -> __main__.A'
+reveal_type(B.new) # E: Revealed type is 'def (factory: def (__main__.B) -> __main__.B) -> __main__.B'
+
+[builtins fixtures/classmethod.pyi]
+
+[case testSelfTypeBound]
+from typing import TypeVar, Callable, cast
+ 
+T = TypeVar('T')
+
+class A:
+    def copy(self: T) -> T:
+        ...
+
+class C(A):    
+    def copy(self: C) -> C:
```

Comment:
It would be nice to have `# Error` markers on lines that the output refers to, as counting lines manually is inconvenient. (`# E:` doesn't work directly, though I think there's now a way to enable less verbose context for error messages and then `# E: ...` should be good.)